docs/madr/decisions/006-mesh-traffic-permission.md

```diff
@@ -0,0 +1,503 @@
+# MeshTrafficPermission
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4222
+
+## Context and Problem Statement
+
+[New policy matching MADR](./005-policy-matching.md) introduces a new approach how Kuma applies configuration to proxies.
+Rolling out strategy implies creating a new set of policies that satisfy a new policy matching.
+Current MADR aims to define a MeshTrafficPermission.
+
+## Considered Options
+
+* Create a MeshTrafficPermission policy
+
+## Decision Outcome
+
+Chosen option: create a MeshTrafficPermission policy
+
+### Overview
+
+MeshTrafficPermission allows blocking unwanted incoming traffic for the service.
+There are might be several reasons why it could be useful to block unwanted traffic:
+* following the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
+* cut off access for clients that shouldn't use your service (legacy clients, clients from another zone, etc.)
+
+### Specification
+
+Top-level targetRef can have all available kinds:
+
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+  name: ...
+```
+
+MeshTrafficPermission is an inbound policy that's why it has a "from" array:
+
+```yaml
+from:
+  - targetRef:
+      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+      name: ...
+```
+
+Configuration for MeshTrafficPermission has a single field "action" with 2 values "ALLOW" or "DENY".
+So the overall specification looks like this:
+
+```yaml
+type: MeshTrafficPermission
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+    name: ...
+  from:
+    - targetRef:
```

Comment:
how do we handle cross-mesh permissions?

```
fromMesh: other-mesh
from:
- kind: MeshService
  name: service-name
```
?

Reply:
Are there cases when you want to combine regular permissions with cross-mesh permissions in one policy? Maybe `MeshService` targetRef should have a `mesh` field, which is empty by default. But if you set it, then it's the service from another mesh:
```yaml
from:
  - kind: MeshService
    mesh: other-mesh
    name: service-name
  - kind: MeshService
    name: backend
```

Reply:
I don't think anyone would _need_ that option. You can always just create two policies I think. The `targetRef` option is there of course, but maybe it's cleaner to add it to the policy spec since I think it only works for `MeshTrafficPermission`.

Reply:
What happens when you concatenate `from` arrays from different levels? Do you have separate `from` arrays for cross mesh and regular permissions?

Reply:
I think we can just treat it internally as if there _were_ a `mesh` field in the `targetRef` entries.

Then again, this is really only a thing for `targetRef` kinds that aren't real objects so maybe it should just be a field for those kinds and we just use validation to enforce these rules.

Reply:
We can create a `mesh` field in the "from" item:
```yaml
type: MeshTrafficPermission
mesh: expose
name: cross-mesh-permission
spec:
  targetRef:
    kind: MeshGatewayRoute
    name: cross-mesh-route
  from:
    - mesh: consume
      targetRef:
        kind: MeshService
        name: backend
      conf:
        action: ALLOW
    - mesh: consume
      targetRef:
        kind: MeshServiceSubset
        name: backend
        tags:
          version: v1
      conf:
        action: DENY
```
so only MeshTrafficPermission will have `mesh`. Other inbound policies with "from" arrays (FaultInjection or RateLimit) won't have it

Reply:
Ah of course! Nice

Reply:
Who is doing a great job on the team? What have they done?

Reply:
What's the reason to have this at the same level as `targetRef` rather than being a field in the `targetRef`? I'm not sure I understand why we gave up on this

Reply:
Added "mesh" into targetRef

Reply:
@lahabana you mean gave up on it being a field in `targetRef`? I was just concerned that it may not be applicable/possible for every policy but putting it in `targetRef` would apply to every policy. But maybe it is possible for every policy or eventually will be?